arangod/Pregel/Algos/HITSKleinberg.cpp

```diff
@@ -0,0 +1,331 @@
+
+        aggregate<double>(hubAggregator, nonNormalizedHub * nonNormalizedHub);
+
+        // normalize auth (hubDivisor is not ready yet, cannot normalize hubs)
+        const auto nonNormalizedAuth = mutableVertexData()->nonNormalizedAuth;
+        const auto normalizedUpdatedAuth = nonNormalizedAuth / ctx->authDivisor;
+        const auto diff =
+            fabs(mutableVertexData()->normalizedAuth - normalizedUpdatedAuth);
+        mutableVertexData()->normalizedAuth = normalizedUpdatedAuth;
+
+        aggregate<double>(maxDiffAuthAggregator, diff);
+        sendHubToOutNeighbors(nonNormalizedHub);
+
+        // authorities and hubs are updated to the same iteration,
+        // authorities are normalized, hubs are not normalized
+
+        // Next state: updateAuthNormalizeHub or finallyNormalizeHubs
+        break;
+      }
+
+      case State::updateAuthNormalizeHub: {
+        // authorities and hubs are updated to the same iteration,
+        // authorities are normalized, hubs are not normalized
+        updateStoreAndSendAuth(messages, ctx->hubDivisor);
+        const auto nonNormalizedHub = mutableVertexData()->nonNormalizedHub;
+        const auto normalizedUpdatedHub = nonNormalizedHub / ctx->hubDivisor;
+        const auto diff =
+            fabs(mutableVertexData()->normalizedHub - normalizedUpdatedHub);
+        mutableVertexData()->normalizedHub = normalizedUpdatedHub;
+        aggregate<double>(maxDiffHubAggregator, diff);
+
+        // authorities are updated one iteration more than hubs,
+        // authorities are not normalized, hubs are normalized
+
+        // next state: updateHubNormalizeAuth
+        break;
+      }
+
+      case State::finallyNormalizeHubs:
+        // authorities and hubs are updated to the same iteration,
+        // authorities are normalized, hubs are not normalized
+        // last iteration
+        mutableVertexData()->normalizedHub =
+            mutableVertexData()->nonNormalizedHub / ctx->hubDivisor;
+        break;
+
+      case State::finallyNormalizeAuths:
+        // authorities and hubs are updated to the same iteration,
+        // authorities are not normalized, hubs are normalized
+        // last iteration
+        mutableVertexData()->normalizedAuth =
+            mutableVertexData()->nonNormalizedAuth / ctx->authDivisor;
+        break;
+    }
+  }
+
+ private:
+  // At the beginning, we don't have differences between the current and the
+  // previous values yet. If we don't report any difference, the default
+  // difference 0 will be taken and the process terminates.
+  void reportFakeDifference() {
+    // so that 0 != diff < epsilon
+    aggregate(maxDiffAuthAggregator, epsilon + 1.0);
+  }
+
+  void updateStoreAndSendAuth(
+      MessageIterator<SenderMessage<double>> const& messages,
+      double hubDivisor) {
+    // compute the new auth
+    double auth = 0.0;
+    for (SenderMessage<double> const* message : messages) {
```

Comment:
I am getting a bit nervous about `nullptr` in these; I realise there's no good solution to this problem (we'll address it in refactor) you could put a TRI_ASSERT in :shrug:

Reply:
Done